python/kotekan/shared_memory_buffer.py

```diff
@@ -0,0 +1,539 @@
+        self._last_access_record[times, :] = access_record_after_copy[times, :]
+
+        # return visRaw()
+        return self._return_data_copy_since(timestamp)
+
+    def _max_ts_per_freq(self, access_record):
+        # get the most recent timestamp for each time slot
+        last_ts = [None] * self.num_time
+        for t in range(self.num_time):
+            for f in range(self.num_freq):
+                if (
+                    last_ts[t] is None and access_record[t][f] != self.invalid_value
+                ) or (last_ts[t] is None or access_record[t][f] > last_ts[t]):
+                    last_ts[t] = access_record[t][f]
+        return last_ts
+
+    def _sort_timestamps(self, timestamps):
+        timestamp_sorted, idxs = list(
+            zip(
+                *sorted(
+                    [
+                        (val, i)
+                        for i, val in enumerate(timestamps)
+                        if val != self.invalid_value
+                    ]
+                )
+            )
+        )
+        return timestamp_sorted, idxs
+
+    def _check_for_identical_timestamps(self, timestmaps):
+        # there should not be multiple identical entries
+        if len(timestmaps) != len(set(timestmaps)):
+            raise SharedMemoryError(
+                "Found duplicate timestamps in access record: {}".format(timestmaps)
+            )
+
+    def _filter_since(self, access_record, timestamp):
+        last_ts = self._max_ts_per_freq(access_record)
+
+        # remove invalid timestmaps
+        while True:
+            try:
+                last_ts.remove(self.invalid_value)
+            except ValueError:
+                break
+
+        self._check_for_identical_timestamps(last_ts)
+
+        # return timestamps newer than the given one
+        ll = [i for i, v in enumerate(last_ts) if v > timestamp]
+        logger.debug(
+            "Found {} time slots that are newer than {} in {}: {}.".format(
+                len(ll), timestamp, last_ts, ll
+            )
+        )
+        return ll
+
+    def _return_data_copy_since(self, timestamp):
+        if not self._time_index_map:
+            return None
+
+        # sort time index map
+        last_ts, idxs = list(zip(*sorted(self._time_index_map.items())))
+
+        # get indexes of timestamps newer than the given on
+        idxs = [i for i, v in zip(idxs, last_ts) if v > timestamp]
+        logger.debug(
+            "Found {} time slots that are newer than {} in {}: {}.".format(
+                len(idxs), timestamp, last_ts, idxs
+            )
+        )
+
+        # select last n time slots
+        return VisRaw.from_nparray(
+            self._data[idxs, :],
+            self.size_frame,
+            len(idxs),
+            self.num_freq,
+            num_elements=7,
+            num_stack=28,
+            num_ev=0,
+        )
+
+    def _copy_from_shm(self, times, access_record):
+        # copy data updates within the last n time slots
```

Comment:
This doesn't quite do what the comment suggests. It copies out the data in the specified slots.